Add sorting function to sortable table

holdenhinkle opened the pull request:
## Description
Closes <https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/230>

The following functionality was added to the Table component. Includes new unit tests.
-  Basic sorting functionality is added to component
-  Sorting function can be overridden by passing in a prop
-  Unit testing is added to confirm sort ascending and descending behavior
-  Sorting behavior should be defined for String, Number, and maybe Boolean
-  Create and/or modify unit tests for this component to include an aXe scan

## PR for CSS Update
The PR for CSS updates is here - department-of-veterans-affairs/vets-website#18726

## Testing done
New unit tests were added to test the sort function.

A new aXe check unit test was added to check the table with a sort column.

## Screenshots
See below.

## Acceptance criteria
See above.

## Definition of done
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)

ddzz reviewed (changes requested):
There are border formatting issues with the sortable column header:

<img width="859" alt="Screen Shot 2021-09-10 at 4 34 16 PM" src="https://user-images.githubusercontent.com/3535749/132920104-ee825a9b-9374-4b3f-bb89-b022f88a4626.png">

holdenhinkle commented:
> There are border formatting issues with the sortable column header:
> 
> <img alt="Screen Shot 2021-09-10 at 4 34 16 PM" width="859" src="https://user-images.githubusercontent.com/3535749/132920104-ee825a9b-9374-4b3f-bb89-b022f88a4626.png">

@ddzz I thought that was part of the design, to make the sortable column stand out more. I could be wrong. That's how Storybook had it before I started on this ticket. I'll find out on Monday.

holdenhinkle commented:
@pjhill In `src/components/Table/Table.jsx` someone explicity removed the top, right and left borders in the `th` element for sortable columns:

```
const borderClasses =
  'vads-u-border-top--0 vads-u-border-right--0 vads-u-border-left--0 vads-u-font-family--sans vads-u-padding--0 vads-u-padding-y--0p5 medium-screen:vads-u-padding--2';
```

If I remove the `vads-u-border-top--0 vads-u-border-right--0 vads-u-border-left--0` classes, the table header looks 'right'. 

Should I leave it as-is? Or remove those css classes so it looks 'right'? Or ask someone? If I should ask someone, do you know who I should ask?

Thanks!

holdenhinkle commented:
> @pjhill In `src/components/Table/Table.jsx` someone explicity removed the top, right and left borders in the `th` element for sortable columns:
> 
> ```
> const borderClasses =
>   'vads-u-border-top--0 vads-u-border-right--0 vads-u-border-left--0 vads-u-font-family--sans vads-u-padding--0 vads-u-padding-y--0p5 medium-screen:vads-u-padding--2';
> ```
> 
> If I remove the `vads-u-border-top--0 vads-u-border-right--0 vads-u-border-left--0` classes, the table header looks 'right'.
> 
> Should I leave it as-is? Or remove those css classes so it looks 'right'? Or ask someone? If I should ask someone, do you know who I should ask?
> 
> Thanks!

@GnatalieH, what are your thoughts on leaving the top, right and left border `0` on sortable columns like it is now (see the screenshots above)? Or should I make them `1` like the other columns?

GnatalieH commented:
@holdenhinkle thanks for asking, I would make the border on each column "1" so that it's the same everywhere.

holdenhinkle commented:
All `th` elements in `thead` have a border of `1` now:

![image](https://user-images.githubusercontent.com/9326247/133438399-5352256d-52a2-498d-9368-44cd85784f60.png)

ddzz commented:
@holdenhinkle @GnatalieH The latest changes affect the entire table and not just the header. Is this fine?

Before:

<img width="984" alt="Screen Shot 2021-09-15 at 8 37 03 AM" src="https://user-images.githubusercontent.com/3535749/133443529-3b8cd100-e73d-42d0-a9b3-97f7c065f8bd.png">


After:

<img width="978" alt="Screen Shot 2021-09-15 at 8 36 03 AM" src="https://user-images.githubusercontent.com/3535749/133443442-b8c64bff-0b9a-4903-84f0-32919819b9bd.png">

GnatalieH commented:
> @holdenhinkle @GnatalieH The latest changes affect the entire table and not just the header. Is this fine?
> 
> Before:
> 
> <img alt="Screen Shot 2021-09-15 at 8 37 03 AM" width="984" src="https://user-images.githubusercontent.com/3535749/133443529-3b8cd100-e73d-42d0-a9b3-97f7c065f8bd.png">
> 
> After:
> 
> <img alt="Screen Shot 2021-09-15 at 8 36 03 AM" width="978" src="https://user-images.githubusercontent.com/3535749/133443442-b8c64bff-0b9a-4903-84f0-32919819b9bd.png">

@rtwell do you mind the borders on all sides in tables ("after" image)?

holdenhinkle commented:
Hey @ddzz I fixed that. And I also updated the positioning of the FA carets. The PR for CSS updates here - https://github.com/department-of-veterans-affairs/vets-website/pull/18726

![image](https://user-images.githubusercontent.com/9326247/133497755-05e3b393-4487-4d89-b6bf-19edc6acc857.png)

GnatalieH commented:
@holdenhinkle that looks good! Question: have we considered using the arrow icons for sorting that they use on USWDS?

![image](https://user-images.githubusercontent.com/88347129/133499415-1161c06d-1bb6-433c-a0ba-73577d1d6fd3.png)

Sorry if I'm committing scope creep. Aside from style, I would like to make sure the icon we use is focusable no matter what. Thanks.

holdenhinkle commented:
Hi @GnatalieH 
>  Have we considered using the arrow icons for sorting that they use on USWDS?

No, I haven't considered that. The Font Awesome arrows were already being used when I picked this ticket up. Do you want me to use the USWDS arrows instead?

>  I would like to make sure the icon we use is focusable no matter what.

Can you please elaborate on this for me? I ask partly because tables can't be sorted on mobile currently. Also, the button that the icon is inside of is focusable, and you can toggle the direction of a column with your keyboard when it's in focus. But it sounds like you also want the icon itself to be focusable, is that correct? If so, can I ask why? Sorry, I'm not an expert in accessibility. Are you hoping the icon would tell the reader with the column is being sorted ASC or DESC? Thanks!

holdenhinkle commented:
@GnatalieH 

Okay, I swapped out the Font Awesome arrows with USWDS arrows:
![image](https://user-images.githubusercontent.com/9326247/133509862-8a8bca78-70ac-4d2b-bce8-72cd77eab17e.png)

The button that the icons are in is focusable which allows you to toggle the sort direction, but the icons themselves aren't focusable. Is that okay?

holdenhinkle commented:
I added `aria-label`s for the sort button:
- 'sort data by ascending'
- 'sort data by descending'

GnatalieH commented:
@holdenhinkle as long as the button the icon is in is focusable, I think that's sufficient. I mostly wanted to make sure that the sort function can be accessed via keyboard. I like the big arrow icons! Thanks for your work on this.

GnatalieH commented:
@holdenhinkle do you know if this sorting pattern has passed all relevant accessibility testing? If so, I'd like to get this into the design system.

holdenhinkle commented:
@noahgelman , can you please check this for accessibility? Thanks!

noahgelman reviewed (approved):
LGTM!

holdenhinkle commented:
> @holdenhinkle do you know if this sorting pattern has passed all relevant accessibility testing? If so, I'd like to get this into the design system.

@GnatalieH Noah just gave it the thumbs up :-)

GnatalieH commented:
> > @holdenhinkle do you know if this sorting pattern has passed all relevant accessibility testing? If so, I'd like to get this into the design system.
> 
> @GnatalieH Noah just gave it the thumbs up :-)

Sweet! Thanks @holdenhinkle and @noahgelman!

ddzz commented:
@holdenhinkle Is there a CSS issue with the arrow here?
<img width="969" alt="Screen Shot 2021-09-21 at 9 27 07 AM" src="https://user-images.githubusercontent.com/3535749/134189632-540e1a38-82db-429c-800f-050efff2b3e7.png">

holdenhinkle commented:
> @holdenhinkle Is there a CSS issue with the arrow here?
> 
> <img alt="Screen Shot 2021-09-21 at 9 27 07 AM" width="969" src="https://user-images.githubusercontent.com/3535749/134189632-540e1a38-82db-429c-800f-050efff2b3e7.png">

There's a PR for the CSS update in `vets-website` here - https://github.com/department-of-veterans-affairs/vets-website/pull/18726

holdenhinkle commented:
@GnatalieH If you're good with this PR, can you please approve it? Thanks!

GnatalieH commented:
Hi. The arrow should be to the right of the text, like this image:

![image](https://user-images.githubusercontent.com/88347129/134357273-f6fc15b8-7ea7-4e4a-94fd-d64a46eaa3fa.png)

Otherwise this pattern looks out of sync with standard column sort patterns. The sort icon is always to the left or right of the text, in my observation. Thanks.

holdenhinkle commented:
> Hi @holdenhinkle and @ddzz. The arrow should be to the right of the text, like this image:
> 
> ![image](https://user-images.githubusercontent.com/88347129/134357273-f6fc15b8-7ea7-4e4a-94fd-d64a46eaa3fa.png)
> 
> Otherwise this pattern looks out of sync with standard column sort patterns. The sort icon is always to the left or right of the text, in my observation. Thanks.
> 
> PS: can you confirm that the column headers are using Helvetica? Thanks, again!

@GnatalieH The CSS that powers this is in `vets-website` and here's the PR to update the CSS so the arrow is on the right - department-of-veterans-affairs/vets-website#18726

GnatalieH commented:
@holdenhinkle I met with Chris and Brooks this morning about this issue. I approve the design of the sort pattern _if_ the arrow is going to be right-aligned in regards to the column header text label. @bkjohnson will need to comment on the implementation, however.

bkjohnson reviewed (changes requested):
CSS changes for components in `component-library` should not be made in `vets-website`. There are other projects besides `vets-website` that use our component library.

https://github.com/department-of-veterans-affairs/vets-website/pull/18726#discussion_r714379165

bkjohnson commented on src/components/Table/Table.jsx:
```diff
@@ -27,41 +27,79 @@ const cellProps = (
+  let sortLabel;
+  let rowData;
+
+  useEffect(() => {
+    rowData = sortData();
+    sortLabel = getSortLabel();
+  });
```
I'm not sure why these are being assigned inside an effect. Can you explain why?

holdenhinkle replied:
When the `sortDirection` state is changed, then we re-sort the the data and set the `aria` sort label.

Actually, I can remove the `sortLabel` stuff.

holdenhinkle commented on src/components/Table/Table.jsx:
```diff
@@ -27,41 +27,79 @@ const cellProps = (
+                <button
+                  className="va-button-link vads-u-font-weight--bold vads-u-color--base vads-u-text-decoration--none sort"
+                  onClick={() =>
+                    setSortDirection(sortDirection === 'ASC' ? 'DESC' : 'ASC')
```
It needs to be the way it is--`setSortDirection(sortDirection === 'ASC' ? 'DESC' : 'ASC')`--because we're setting `sortDirection` to the opposite of whatever it is when the user clicks the sort button.

Also, if no column is set to be sorted, then `currentSort.order` is `undefined`. If it's `undefined`, then we don't sort any of the data. So that also need to stay the way it is, with the current implementation.

bkjohnson replied:
Ahh, the opposite makes sense. Somehow I overlooked that.

holdenhinkle commented:
@bkjohnson :
> Also, down in the propTypes there is a JSDoc comment for currentSort that says Currently unused.. This should probably go now.

Done, thanks.
